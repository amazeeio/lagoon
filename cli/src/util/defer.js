// @flow

type Defer<R> = {
  promise: Promise<R>,
  resolve: (result: Promise<R> | R) => void,
  reject: (error: any) => void,
};

/**
 * Creates a defer object, which is useful to combine
 * Promises with nodelike callback interfaces
 */
export default function defer(): Defer<*> {
  const deferred = {};

  deferred.promise = new Promise((resolve, reject) => {
    deferred.resolve = resolve;
    deferred.reject = reject;
  });

  return deferred;
}
