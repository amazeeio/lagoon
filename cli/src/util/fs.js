// @flow

/**
 * This module exists to promisify all the fs functionality
 */

import fs from 'fs';
// Polyfill for Node 8's util.promisify
import promisify from 'util.promisify';

import type { Stats } from 'fs';

type ReadFileFn = (filename: string, enc?: string) => Promise<Buffer | string>;
type WriteFileFn = (
  filename: string,
  data: Buffer | string,
  options?: Object | string,
) => Promise<void>;
type UnlinkFn = (filename: string) => Promise<void>;

export const readFile: ReadFileFn = promisify(fs.readFile);
export const writeFile: WriteFileFn = promisify(fs.writeFile);
export const unlink: UnlinkFn = promisify(fs.unlink);

// TODO: @ryyppy Can we use something simpler like promisify(fs.access)?
export async function fileExists(path: string): Promise<boolean> {
  try {
    const stats: Stats = await promisify(fs.lstat)(path);

    if (stats.isFile()) {
      return true;
    }

    return false;
  } catch (err) {
    return false;
  }
}
