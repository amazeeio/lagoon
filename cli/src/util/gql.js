// @flow

// A no-op tagged template function to signal that queries are GraphQL for tooling
export default function gql(
  literals: Array<string>,
  ...substitutions: Array<any>
) {
  let result = '';

  // run the loop only for the substitution count
  substitutions.forEach((substitution, key) => {
    result += literals[key];
    result += substitution;
  });

  // add the last literal
  result += literals[literals.length - 1];

  return result;
}
