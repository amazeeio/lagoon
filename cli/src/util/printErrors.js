// @flow

import { red } from 'chalk';
import R from 'ramda';

type Cerr = typeof console.error;
type LaguError = { +message: string, +stack?: string };

export function printErrors(
  cerr: Cerr,
  ...errors: Array<LaguError | string>
): number {
  R.compose(
    R.forEach(cerr),
    R.map(red),
    R.map(
      err =>
        typeof err === 'string'
          ? err
          : R.propOr(
            // ...otherwise, use the message property as a fallback
            R.prop('message', err),
            // Try to get the stack...
            'stack',
            err,
          ),
    ),
  )(errors);
  return 1;
}

export function printNoConfigError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      '.lagoon.yml config file not found. Please create one with "lagoon init".\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-init',
  });
}

export function printProjectConfigurationError(cerr: Cerr): number {
  return printErrors(cerr, {
    message:
      'No project configured. Please create a .lagoon.yml config file with "lagoon init" or pass a project to this command via the --project option.\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-init',
  });
}

export function printGraphQLErrors(
  cerr: Cerr,
  ...errors: Array<LaguError>
): number {
  if (
    R.find(R.propEq('message', 'Unauthorized - Bearer Token Required'))(errors)
  ) {
    return printErrors(cerr, {
      message:
        'No authentication token found. Please log in first with "lagoon login".\nOnline documentation: https://github.com/amazeeio/lagoon/blob/master/cli/README.md#lagoon-login',
    });
  }
  const errorsPluralized = R.length(errors) === 1 ? 'an error' : 'errors';
  return printErrors(
    cerr,
    `Oops! The Lagoon API returned ${errorsPluralized}:`,
    ...errors,
  );
}
