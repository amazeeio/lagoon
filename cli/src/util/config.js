// @flow

import { fileExists, writeFile } from './fs';
import yaml from 'js-yaml';

import path from 'path';
import co from 'co';

type DeployTask = {
  before_deploy: Array<string>,
  after_deploy: Array<string>,
};

// TODO: Type the rest of the config
export type AmazeeConfig = {
  sitegroup: string,
  deploy_tasks: {
    [name: string]: DeployTask,
  },
};

export type AmazeeConfigInput = {
  sitegroup: string,
};

export function createConfig(
  filepath: string,
  inputOptions: AmazeeConfigInput,
): Promise<void> {
  const config: AmazeeConfig = {
    ...inputOptions,
    deploy_tasks: {
      task1: {
        before_deploy: [],
        after_deploy: [],
      },
    },
  };
  const yamlConfig = yaml.safeDump(config);
  return writeFile(filepath, yamlConfig);
}

function* walker(
  dir: string,
  filename: string,
  root: string,
): Generator<*, ?string, *> {
  let next = dir;

  while (true) {
    const full = path.join(next, filename);
    const exists = yield fileExists(full);

    if (exists) {
      return full;
    } else if (next === root) {
      // ${root}/${filename} does not exist, quit
      return null;
    }

    // Otherwise go up one level
    next = path.dirname(next);
  }
}

/**
 * Tries to resolve a given filename, starting at cwd.
 * Will try ${cwd}/$filename, if not found it will step up
 * one directory (..) until it reaches root (/${filename}).
 *
 * If no file was found, it will return null, otherwise the path
 * of the found file.
 */
export async function findConfig(
  filename: string,
  cwd: string,
): Promise<?string> {
  const start = path.join(cwd, filename);
  const { root } = path.parse(start);

  return co(walker(cwd, filename, root));
}

export function parseConfig(yamlContent: string): AmazeeConfig {
  // TODO: Add schema validation in there if necessary
  return yaml.safeLoad(yamlContent);
}
