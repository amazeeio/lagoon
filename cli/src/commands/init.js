// @flow

import path from 'path';
import { green } from 'chalk';
import R from 'ramda';
import inquirer from 'inquirer';
import createConfig from '../createConfig';
import { fileExists } from '../util/fs';
import { printErrors } from '../printErrors';

import typeof Yargs from 'yargs';
import type { BaseArgs } from './index';

const name = 'init';
const description =
  'Create a .amazeeio.yml config file in the current working directory';

type GetOverwriteOptionArgs = {
  exists: boolean,
  filepath: string,
  overwriteOption?: boolean,
};

const getOverwriteOption = async (
  args: GetOverwriteOptionArgs,
): Promise<boolean> =>
  R.cond([
    // If the file doesn't exist, the file doesn't need to be overwritten
    [R.propEq('exists', false), R.F],
    // If the overwrite option for the command has been specified, use the value of that
    [
      R.propSatisfies(
        // Option is not null or undefined
        R.complement(R.isNil),
        'overwriteOption',
      ),
      R.prop('overwriteOption'),
    ],
    // If none of the previous conditions have been satisfied, ask the user if they want to overwrite the file
    [
      R.T,
      async ({ filepath }) => {
        const { overwrite } = await inquirer.prompt([
          {
            type: 'confirm',
            name: 'overwrite',
            message: `File '${filepath}' already exists! Overwrite?`,
            default: false,
          },
        ]);
        return overwrite;
      },
    ],
  ])(args);

export async function setup(yargs: Yargs): Promise<Object> {
  return yargs
    .usage(`$0 ${name} - ${description}`)
    .options({
      overwrite: {
        describe: 'Overwrite the configuration file if it exists',
        type: 'boolean',
        default: undefined,
      },
      sitegroup: {
        describe: 'Name of sitegroup to configure',
        type: 'string',
        alias: 's',
      },
    })
    .example(
      `$0 ${name}`,
      'Create a config file at ./.amazeeio.yml. This will confirm with the user whether to overwrite the config if it already exists and also prompt for a sitegroup name to add to the config.\n',
    )
    .example(
      `$0 ${name} --overwrite`,
      'Overwrite existing config file (do not confirm with the user).\n',
    )
    .example(
      `$0 ${name} --overwrite false`,
      'Prevent overwriting of existing config file (do not confirm with user).\n',
    )
    .example(
      `$0 ${name} --sitegroup my_sitegroup`,
      'Set sitegroup to "my_sitegroup" (do not prompt the user).\n',
    )
    .example(
      `$0 ${name} -s my_sitegroup`,
      'Short form for setting sitegroup to "my_sitegroup" (do not prompt the user).\n',
    )
    .example(
      `$0 ${name} --overwrite --sitegroup my_sitegroup`,
      'Overwrite existing config files and set sitegroup to "my_sitegroup" (do not confirm with or prompt the user).',
    ).argv;
}

type Args = BaseArgs & {
  overwrite: ?boolean,
  sitegroup: ?string,
};

export async function run({
  cwd,
  overwrite: overwriteOption,
  sitegroup,
  clog,
  cerr,
  }: Args): Promise<number> {
  const filepath = path.join(cwd, '.amazeeio.yml');

  const exists = await fileExists(filepath);

  const overwrite = await getOverwriteOption({
    exists,
    filepath,
    overwriteOption,
  });

  if (exists && !overwrite) {
    return printErrors(cerr, `Not overwriting existing file '${filepath}'.`);
  }

  const configInput = sitegroup
    ? { sitegroup }
    : await inquirer.prompt([
      {
        type: 'input',
        name: 'sitegroup',
        message: 'Enter the name of the sitegroup to configure.',
        validate: input =>
          input ? Boolean(input) : 'Please enter a sitegroup.',
      },
    ]);

  try {
    clog(`Creating file '${filepath}'...`);
    await createConfig(filepath, configInput);
    clog(green('Configuration file created!'));
  } catch (e) {
    return printErrors(cerr, `Error occurred while writing to ${filepath}:`, e);
  }

  return 0;
}

export default {
  setup,
  name,
  description,
  run,
};
