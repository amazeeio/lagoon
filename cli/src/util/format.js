// @flow

import R from 'ramda';
import {
  FORMAT,
  FORMAT_CHOICE_TABLE,
  FORMAT_CHOICE_SIMPLE,
  FORMAT_CHOICE_JSON,
  FORMAT_CHOICE_CSV,
} from '../config/globalOptions';
import { getConfig } from '../config';

import type { FormatData } from '../types/Format';

export default function format(data: FormatData) {
  return R.cond([
    [
      R.equals(FORMAT_CHOICE_TABLE),
      () => require('../formatters/table').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_SIMPLE),
      () => require('../formatters/simple').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_JSON),
      () => require('../formatters/json').default(data),
    ],
    [
      R.equals(FORMAT_CHOICE_CSV),
      () => require('../formatters/csv').default(data),
    ],
  ])(R.prop(FORMAT, getConfig()));
}
