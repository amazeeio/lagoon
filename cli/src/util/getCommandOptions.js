// @flow

import R from 'ramda';
import { globalOptions } from '../config/globalOptions';

import type { Config } from '../types/Config';

export function getCommandOptions<CommandOptionsT: { [key: string]: any }>({
  config,
  commandOptions,
}: {|
  config: Config,
  commandOptions: CommandOptionsT,
|}): { [key: $Keys<CommandOptionsT>]: any, ...$Exact<Config> } {
  // Only return the properties that match the command options
  return R.pick(
    R.concat(R.keys(globalOptions), R.keys(commandOptions)),
    config,
  );
}
