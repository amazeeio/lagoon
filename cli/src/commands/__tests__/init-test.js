// @flow

import { fileExists } from '../../util/fs';
import { run } from '../init';

jest.mock('fs');
jest.mock('../../createConfig');
jest.mock('../../util/fs');

function _mock(fn: any): JestMockFn {
  return fn;
}

describe('run', () => {
  // TODO: Also write a test where it bails
  it('should overwrite existing config file if overwrite option passed as true', async () => {
    _mock(fileExists).mockImplementationOnce(() => Promise.resolve(true));

    const clog = jest.fn();
    const cerr = jest.fn();
    const cwd = 'some/path';

    const code = await run({
      config: null,
      _: [],
      $0: '',
      cwd,
      clog,
      cerr,
      overwrite: true,
      sitegroup: 'test',
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });

  it('should not overwrite config file when `overwrite` option set to false', async () => {
    _mock(fileExists).mockImplementationOnce(() => Promise.resolve(true));

    const clog = jest.fn();
    const cerr = jest.fn();
    const cwd = 'some/path';

    const code = await run({
      config: null,
      _: [],
      $0: '',
      cwd,
      clog,
      cerr,
      overwrite: false,
      sitegroup: 'test',
    });

    expect(code).toBe(1);
    expect(cerr.mock.calls).toMatchSnapshot();
  });

  it('should write default yaml to given cwd + .amazeeio.yml', async () => {
    _mock(fileExists).mockImplementationOnce(() => Promise.resolve(false));

    const clog = jest.fn();
    const cerr = jest.fn();
    const cwd = 'some/path';

    const code = await run({
      config: null,
      _: [],
      $0: '',
      cwd,
      clog,
      cerr,
      overwrite: false,
      sitegroup: 'test',
    });

    expect(code).toBe(0);
    expect(clog.mock.calls).toMatchSnapshot();
  });
});
