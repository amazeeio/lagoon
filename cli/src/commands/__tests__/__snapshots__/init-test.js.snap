// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run should not overwrite config file when \`overwrite\` option set to false 1`] = `
Array [
  Array [
    "[31mNot overwriting existing file 'some/path/.amazeeio.yml'.[39m",
  ],
]
`;

exports[`run should overwrite existing config file if overwrite option passed as true 1`] = `
Array [
  Array [
    "Creating file 'some/path/.amazeeio.yml'...",
  ],
  Array [
    "[32mConfiguration file created![39m",
  ],
]
`;

exports[`run should write default yaml to given cwd + .amazeeio.yml 1`] = `
Array [
  Array [
    "Creating file 'some/path/.amazeeio.yml'...",
  ],
  Array [
    "[32mConfiguration file created![39m",
  ],
]
`;
